### ISSUE TITLE
Permanent playlist overwrites and erasures should not be so aggressive and transparent.

### ISSUE THREAD

**Issue by will-ca** [quodlibet/quodlibet#3131]
The proposed change
-------------------

If you click any of the tabs to sort your view by Title, Artist, Album (etc.) while viewing a playlist, the displayed order of songs will be changed with no clear way to revert it to the original playlist's order, and if you then click and drag any tracks to try to fix it, the order of the playlist itself and all associated files is promptly also changed permanently to the sorted order, even though it's not immediately clear and there is no indication or precedent that doing so will affect anything other than the displayed order.

Additionally, if a file is missing or has been renamed, it just gets deleted from the playlist with no confirmation. (A *notification* shows up, but by then it's too late.)

And if the music library's directory has been relocated or renamed, *all* files from *all* playlists get erased permanently from all playlist files in `~/.quodlibet`, with no backups, and with neither confirmation nor indication communicating such a large change. (I just lost multiple playlists, including one with over four hundred songs, because of this.)

None of this should happen, and especially not so easily.

Benefits of this change
-----------------------

Playlists are a great way of organizing your music library. They can be used to tell a story by connecting songs in a sequence, they can help with sorting and finding songs based off of mood or social setting, and they can bear significant sentimental significance as well when either they or their constituent songs are connected to personally important events.

In addition to all this— or perhaps because of it— they can also represent a significant amount of work spent by the user in manually creating and maintaining them.

*Quod Libet should not Irreversibly delete or modify user-created and heavily curated data without any kind of clear prompt.*

Some proposed solutions
---------------------------

The primary potential challenge is in finding an elegant way to change this behaviour.

For the issue with the order being overwritten, I suggest creating a stronger interface-level distinction between "playlist order" and "display order". In every other part of the Quod Libet interface, the way the display is sorted is completely separate from and has no bearing on the actual contents of the material being sorted (like albums' track listings), so the behaviour in the playlist browser is inconsistent, confusing, and troublesome without any explicit indication.

The saved playlist order could be exposed to the user through the creation of a new `~playlistpositions` tag for applicable tracks, generated from `~/.quodlibet/playlists` to parallel the already-existing `~playlists` tag. The value of this tag per each playlist should be displayed in its own column as the default sort key in the Playlists browser.

A new button can then be added to permanently save the playlist's order by updating this tag (along with the appropriate file in `~/.quodlibet/playlists`) to mirror the current display order. Aside from deletions and additions, **the use of this explicit, user-facing control should be the only circumstance where the playlist file on disk would change**.

The user would thus be able to momentarily change the displayed sort order, or even temporarily reorder the tracks for a single listening session, without unwittingly permanently changing the saved copy of the playlist.

When the user would like to deliberately reorder the tracks, the existence of this control element would make the user experience easier as well, since it would indicate the steps needed to update the playlist order whereas such behaviour is currently unexplained anywhere in the interface and has to be deciphered through trial and error. (See #1645.)

This change would make it much easier to avoid unintended operation and loss of data, it would not cause any loss of functionality or significant interface clutter, it would enhance existing functionality as it would provide hints and clarity on the program's behaviour, and it would integrate well with Quod Libet thematically and create opportunities for new functionality as well, since the `~playlistpositions` tag could also be used in other places.

For the issue with tracks being deleted from playlists, I believe a simple confirmation query at startup before making any erasures would work best for me (along with transparently ignoring any still-missing files afterwards). It would introduce no interface clutter and require minimal code changes as a notification dialog box is currently shown anyway, it would solve the problem of deleting the user's data without their permission/intention, and it would make it extremely easy for many users to fix the missing files by simply editing `~/.quodlibet/playlists` or renaming their music library.

**Comment by declension** [comment]
First of all, thanks for the excellent feature request, that's very instructive and echoes things people have said over the years (see my related summary on #708).

> In every other part of the Quod Libet interface, the way the display is sorted is completely separate from and has no bearing on the actual contents of the material being sorted (like albums' track listings), so the behaviour in the playlist browser is inconsistent, confusing, and troublesome without any explicit indication.

_This_ in particular is the problem we have right now. As some background, the assumption throughout the application is / was that:
1. The _songslist_ (set of songs that each browser exposes for viewing its item contents) is shared across these browsers, by and large
2. The songslist is a normal(ish) GTK+ TreeView linking back to a file-based model, with familiar sorting from the headerbar (and filtering from searching)
3. and that there should be no exact duplicates (per filename at least).

The Playlists Browser (as a later entry) fundamentally upset this paradigm as you detailed - it _does_ care about order and it _can_ have duplicates.

Additionally throughout QL / EF, the UI and the library / model / logic code is often tightly coupled, leading to difficult problems / bugs (e.g. #3083) and generally harder rewriting especially for those not very familiar with GTK+.

> The saved playlist order could be exposed to the user through the creation of a new `~playlistpositions` tag for applicable tracks, generated from `~/.quodlibet/playlists` to parallel the already-existing `~playlists` tag. The value of this tag per each playlist should be displayed in its own column as the default sort key in the Playlists browser.

I like this idea, but I'm not sure how it would scale to many playlists. Having a tag per playlist (e.g. `playlistpostions-chilled=3`) seems the most _correct_, but is of course fairly nasty. Munging structured data (e.g. `playlistpositions="{Chilled: 23, Morning Rush: 12, Happy Songs: 34}`" into one tag feels... most nasty.

**Comment by will-ca** [comment]
I personally don't think the second (dictionary-style) tag format is particularly nasty. It does concatenate structured data into a single location, but it preserves the structure of said data in a indexed and decently accessible format, and it contains it all into one tag rather than cluttering the tag list with multiple tags generated from user-created (and potentially problematic) strings.

(E.g.: With the first option, you could get something like `&(artist=Johnny, #(playlistpositions-playlist_with_multiple_capitalization_collisions_and_special_characters_:)_in_its_name<7))`.)

I also don't see how a dictionary-style tag would be that different from the already-existing `~performers` and `~people` tags either, so there is some precedent for concatenating related, structured data into a single object.

**Comment by torotil** [comment]
It seems non-accessible files are also cleaned from playlists when their part of the library that resides on an external device is currently not mounted. For me this makes Quodlibet’s playlists basically unusable. Most of my music collection resides on a SD-card.

I would also consider this part of the feature request a rather major bug. It’s deleting user generated data without warning.

**Comment by declension** [comment]
@torotil sorry about the delay. 

You might find it's there actually, and just needs the directory to be un-hidden in _Preferences_ -> _Library_: 
![Screenshot from 2020-11-28 11-33-04](https://user-images.githubusercontent.com/3322808/100514559-97a06680-316d-11eb-9d6e-cea81c809ee9.png)

**Comment by declension** [comment]
In the search for quick(er) wins here, one improvement might be to try to disable the sortable songlist columns for the playlist browser only.

This would stop the accidental overwrites by users, with the drawback that they are no longer sortable, if people actually _want_ to do that (but I'm searching for use-cases here, seems pretty niche). 

Even then, a _Playlist Sort By_ plugin would be easy and could allow these one-time changes to be done explicitly, if more awkwardly.

**Comment by Meriipu** [comment]
I almost exclusively use the playlist browser and when I do I use the sort columns there regularly.
Unless this is about the floating version of it.

**Comment by declension** [comment]
@Meriipu thanks. And presumably a sorting plugin wouldn't work for you here?

**Comment by Meriipu** [comment]
I sounds a bit more awkward than how it currently works unfortunately.
I use playlists more for their contents than their exact order at creation-time (kind of like a subset of songs).
Manually filtering to a playlist in the searchbar browser is a bit unwieldy.

Perhaps a bandaid-solution might be to when the user is dragging tracks around in a sorted playlist, prompt the user about this action modifying the stored playlists order? Obviously with an option to squelch future such prompts.

**Comment by declension** [comment]
> Perhaps a bandaid-solution might be to when the user is dragging tracks around in a sorted playlist, prompt the user about this action modifying the stored playlists order? Obviously with an option to squelch future such prompts.

Yes, but you'd also have to warn every time a column heading was clicked too, no?

**Comment by Meriipu** [comment]
> > Perhaps a bandaid-solution might be to when the user is dragging tracks around in a sorted playlist, prompt the user about this action modifying the stored playlists order? Obviously with an option to squelch future such prompts.
> 
> Yes, but you'd also have to warn every time a column heading was clicked too, no?

If I click a playlist, order it by a column, click another playlist, then click the first playlist again, the original order is restored.
The order is lost if a track is repositioned after sorting.
